### ISSUE TITLE
Move SGX-specifc attestation ecalls/ocalls to be a part of the attester plugin framework

### ISSUE THREAD

**Issue by jhand2** [openenclave/openenclave#2791]
cc @yentsanglee @shnwc

I have been working to refactor some of our internal EDL files and have some thoughts about the ocalls/ecalls in `common/sgx/edl/sgx.edl` (soon to be `common/sgx/edl/attestation.edl` as of #2784).

These are ocalls such as:
```
oe_get_qetarget_info_ocall()
oe_get_quote_ocall()
...
```

My question is, does it make sense for these to be part of the attester plugin rather than part of the core SDK? Would any other part of the SDK need to call into these? Or are they just for the benefit of the SGX attestation plugin?

**Comment by yentsanglee** [comment]
oe_get_qetarget_info_ocall() and oe_get_quote_ocall() are dependences of **non-plugin** API oe_get_report(). For now, keeping them in sgx.edl (soon to be attestation.edl) is a better option.

**Comment by jhand2** [comment]
Would these ocalls be deprecated by https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation_V3.md?

**Comment by yentsanglee** [comment]
No, they will not be deprecated by CustomAttestation_V3.md. Deep down, SGX verifier plugin and oe_ge_report() share the same code. But oe_get_report() is still a public API that we need to support.

**Comment by jhand2** [comment]
I guess my larger question is in a world where CustomAttestation_v3 is implemented, would oe_get_report still be needed/recommended, or would the new attester function be recommended as a replacement.

In the short term I agree that we should keep these ocalls where they are. I just want to understand the long term state.

**Comment by dthaler** [comment]
> I guess my larger question is in a world where CustomAttestation_v3 is implemented, would oe_get_report still be needed/recommended, or would the new attester function be recommended as a replacement.
> In the short term I agree that we should keep these ocalls where they are. I just want to understand the long term state.

I suspect long term state is probably that oe_get_report is deprecated and replaced with oe_get_evidence.

**Comment by shnwc** [comment]
I think we should do the internal refactoring sooner rather than later, to isolate SGX-specific plugin code and put them into their own enclave and host libraries. So the OE SDK core does not have dependency on SGX specific libraries (DCAP or quote-ex), nor need to define names like `OE_LINK_SGX_DCAP_QL`.

My implementation for the [integration of SGX quote-ex library](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/SGX_QuoteEx_Integration.md) requires change in many enclave- and host-side files (as shown in [my fork](https://github.com/shnwc/openenclave)), and the capability to support multiple plugins in a single library can't fit with the currently-implemented public API. So my implementation can't be upstreamed to the current code base yet. It would be good to do the refactoring along with the integration effort.

Refactoring the OE SDK so that the the OE SDK core does not depend on SGX packages also enables the effort to port SGX SDK to be built with OE SDK. Current implementation has a circular dependency for SGX SDK porting, as building of the OE SDK depends on SGX packages, but building of the SGX packages requires the latest OE SDK.

**Comment by jhand2** [comment]
@shnwc Just to make sure we're on the same page, OE does not support loading shared libraries. So we could (and should) split SGX-specific pieces into a separate .a file, but it will still need to be linked into the enclave binary before signing.

**Comment by shnwc** [comment]
@jhand2 yes, SGX-specific plugin components could include an enclave-side .a file, a host-side .a file, and optionally a host-side .so file. The enclave-side .a files will be linked to OE SDK application enclave projects in their build process before signing.

**Comment by yentsanglee** [comment]
Replacing oe_get_report with oe_get_evidence is the right direction to go. oe_get_report is widely used so deprecating it has to be planned out.

**Comment by shnwc** [comment]
After refactoring, the legacy oe_get_report() API can still be supported, now as part of the SGX-specific plugin library. This API can be implemented so that If the SGX-specific library is not linked to the application enclave project, it would return OE_UNSUPPORTED.

**Comment by CodeMonkeyLeet** [comment]
Per SIG-arch meeting, deferring this to SIG-attestation.
